diff --git a/data/zsh-completion.in b/data/zsh-completion.in
index fe621cb..bbe9d07 100644
--- a/data/zsh-completion.in
+++ b/data/zsh-completion.in
@@ -26,21 +26,21 @@ _arguments -C -s -S "$all_opts[@]" && ret=0
 
 local PLUGIN
 local -a exts
-for PLUGIN in @PLUGINDIR@/lib*.so; do
+for PLUGIN in @PLUGINDIR@/lib*.dylib; do
   case ${PLUGIN##*/} in
-    libpdf-poppler.so)
+    libpdf-poppler.dylib)
       exts+=( pdf PDF )
       ;;
-    libpdf-mupdf.so)
+    libpdf-mupdf.dylib)
       exts+=( pdf PDF epub oxps )
       ;;
-    libps.so)
+    libps.dylib)
       exts+=( ps eps epsi epsf )
       ;;
-    libdjvu.so)
+    libdjvu.dylib)
       exts+=( djvu djv )
       ;;
-    libcb.so)
+    libcb.dylib)
       exts+=( cb7 cbr cbz cbt rar zip 7z tar )
       ;;
   esac
diff --git a/meson.build b/meson.build
index 65a9194..f475417 100644
--- a/meson.build
+++ b/meson.build
@@ -74,7 +74,7 @@ defines = [
 ]
 
 if host_machine.system() == 'darwin'
-  defines += '-DGTKOSXAPPLICATION'
+  defines += '-DMAC_INTEGRATION'
 endif
 
 # compile flags
@@ -164,6 +164,15 @@ headers = files(
 )
 headers += version_header
 
+if host_machine.system() == 'darwin'
+  sources += files(
+    'zathura/mac-integration.c',
+  )
+  sources += files(
+    'zathura/mac-integration.h',
+  )
+endif
+
 # zathura helper library
 libzathura = static_library('zathura',
   sources + ['zathura/database-sqlite.c'],
diff --git a/zathura/mac-integration.c b/zathura/mac-integration.c
new file mode 100644
index 0000000..5ff15fc
--- /dev/null
+++ b/zathura/mac-integration.c
@@ -0,0 +1,75 @@
+/* SPDX-License-Identifier: Zlib */
+
+#include <girara/log.h>
+#include <glib.h>
+#include <gtkosxapplication.h>
+
+#include "mac-integration.h"
+
+#include "plugin.h"
+#include "shortcuts.h"
+#include "zathura.h"
+
+
+static gboolean is_inside_bundle(void);
+static void load_plugins_from_bundle(zathura_t* zathura);
+static gboolean openfile_callback(GtkosxApplication* app, gchar* path, zathura_t* zathura);
+
+
+void zathura_mac_init(zathura_t* zathura) {
+  if (is_inside_bundle()) {
+    load_plugins_from_bundle(zathura);
+  }
+
+  GtkosxApplication* zathura_mac = g_object_new(GTKOSX_TYPE_APPLICATION, NULL);
+  gtkosx_application_set_use_quartz_accelerators(zathura_mac, FALSE);
+  gtkosx_application_ready(zathura_mac);
+
+  g_signal_connect(G_OBJECT(zathura_mac), "NSApplicationOpenFile", G_CALLBACK(openfile_callback), zathura);
+}
+
+static gboolean is_inside_bundle(void) {
+  gchar* id = gtkosx_application_get_bundle_id();
+  gboolean result = id != NULL;
+  g_free(id);
+  return result;
+}
+
+static void load_plugins_from_bundle(zathura_t* zathura) {
+  zathura_plugin_manager_t* plugin_manager = zathura->plugins.manager;
+
+  gchar* resource_path = gtkosx_application_get_resource_path();
+  gchar* plugins_path = g_build_path("/", resource_path, "plugins", NULL);
+
+  zathura_plugin_manager_set_dir(plugin_manager, plugins_path);
+  zathura_plugin_manager_load(plugin_manager);
+
+  g_free(resource_path);
+  g_free(plugins_path);
+}
+
+static gboolean openfile_callback(GtkosxApplication* UNUSED(app), gchar* path, zathura_t* zathura) {
+  if (zathura_has_document(zathura) == false) {
+    document_open_idle(zathura, path, NULL, ZATHURA_PAGE_NUMBER_UNSPECIFIED, NULL, NULL, NULL, NULL);
+    return true;
+  }
+
+  pid_t pid = fork();
+
+  if (pid == 0) { /* child */
+    if (setsid() == -1) {
+      girara_error("Could not start new process group: %s", strerror(errno));
+    }
+
+    char *args[] = { gtkosx_application_get_executable_path(), path, NULL };
+    execv(args[0], args);
+
+    girara_error("Could not replaces process image: %s", strerror(errno));
+    exit(1);
+  } else if (pid < 0) { /* error */
+    girara_error("Could not fork: %s", strerror(errno));
+    /* Should we somehow notify user about this situation? */
+  }
+
+  return true;
+}
diff --git a/zathura/mac-integration.h b/zathura/mac-integration.h
new file mode 100644
index 0000000..07365e6
--- /dev/null
+++ b/zathura/mac-integration.h
@@ -0,0 +1,15 @@
+/* SPDX-License-Identifier: Zlib */
+
+#ifndef MAC_INTEGRATION_H
+#define MAC_INTEGRATION_H
+
+#include "zathura.h"
+
+/**
+ * Initializes zathura OSX integration
+ *
+ * @param zathura The zathura session
+ */
+void zathura_mac_init(zathura_t* zathura);
+
+#endif /* MAC_INTEGRATION_H */
diff --git a/zathura/main.c b/zathura/main.c
index a439a78..7377272 100644
--- a/zathura/main.c
+++ b/zathura/main.c
@@ -1,7 +1,7 @@
 /* SPDX-License-Identifier: Zlib */
 
-#ifdef GTKOSXAPPLICATION
-#include <gtkosxapplication.h>
+#ifdef MAC_INTEGRATION
+#include "mac-integration.h"
 #endif
 
 #include <girara-gtk/settings.h>
@@ -283,17 +283,9 @@ GIRARA_VISIBLE int main(int argc, char* argv[]) {
     return -1;
   }
 
-#ifdef GTKOSXAPPLICATION
-  GtkosxApplication* zathuraApp = g_object_new(GTKOSX_TYPE_APPLICATION, NULL);
-  gtkosx_application_set_use_quartz_accelerators(zathuraApp, FALSE);
-  gtkosx_application_ready(zathuraApp);
-  {
-    const gchar* id = gtkosx_application_get_bundle_id();
-    if (id != NULL) {
-      girara_warning("TestIntegration Error! Bundle has ID %s", id);
-    }
-  }
-#endif // GTKOSXAPPLICATION
+#ifdef MAC_INTEGRATION
+  zathura_mac_init(zathura);
+#endif
 
   /* run zathura */
   gtk_main();
